lib/clients.ts

```diff
@@ -0,0 +1,113 @@
+import { Consumer, Producer, Kafka, EachMessagePayload, logLevel } from 'kafkajs';
+
+class KafkaClient {
+  bootstrapServer: string;
+  username: string;
+  password: string;
+  kafka: Kafka;
+
+  constructor(bootstrapServer: string, username: string, password: string) {
+    this.bootstrapServer = bootstrapServer;
+    this.username = username;
+    this.password = password;
+    this.kafka = new Kafka({
+      brokers: [bootstrapServer],
+      ssl: true,
+      sasl: {
+        // TODO add oauth
+        mechanism: 'plain',
+        username: username,
+        password: password
+      },
+      logLevel: logLevel.INFO
+    });
+  }
+}
+
+export class KafkaProducer extends KafkaClient {
+  private producer: Producer;
+
+  constructor(bootstrapServer: string, username: string, password: string) {
+    super(bootstrapServer, username, password);
+    this.producer = this.kafka.producer();
+  }
+
+  async produceMessages(topicName: string, messageCount: number, key: string) {
+    await this.producer.connect();
+
+    const messages = this.generateMessages(messageCount, key);
+    return await this.producer
+      .send({
+        topic: topicName,
+        messages: messages,
+        acks: -1
+      })
+      .then((response) => {
+        this.kafka.logger().info(`Messages sent #${response}`, {
+          response,
+          messageCount
+        });
+        return true;
+      })
+      .catch((e) => {
+        this.kafka.logger().error(`[${KafkaProducer.name}] ${e.message}`, { stack: e.stack });
+        return e;
+      });
+  }
+
+  private generateMessage(key: string) {
+    return {
+      key: `key-${key}`,
+      value: `value-${key}-${new Date().toISOString()}`
+    };
+  }
+
+  private generateMessages(messageCount: number, key: string) {
+    return Array(messageCount).fill(this.generateMessage(key));
+  }
+}
+
+export class KafkaConsumer extends KafkaClient {
+  private kafkaConsumer: Consumer;
+
+  constructor(bootstrapServer: string, consumerGroup: string, username: string, password: string) {
+    super(bootstrapServer, username, password);
+    this.kafkaConsumer = this.kafka.consumer({ groupId: consumerGroup });
+  }
+
+  public async consumeMessages(topic: string, expectedMsgCount: number, fromBeginning = true): Promise<boolean> {
+    let msgCount = 0;
```

Comment:
I am not sure if retries make sense. In case the consumer is attached it will eventually consume messages if the messages will be available. In case the consumer won't be able to read from the topic, it will fail anyway. We are not even able to say "okay, consume just 200 messages" with some simple implementation as we do in Java for example. 

Or did you mean retries for consuming the same messages again?

Reply:
fair answer 👍 
let's evolve the code when needed.